### PR TITLE
ISPN-11151 - Migrating some remote tests from jdg-functional-tests to upstream

### PR DESCRIPTION
**Pull request by gustavolira**
https://issues.redhat.com/browse/ISPN-11151

### CODE REVIEW

**Comment by gustavolira** [comment]
@ryanemerson I'm just reopening #8162

**Comment by ryanemerson** [comment]
@gustavolira It seems like your changes aren't popular with the `ThreadLeakChecker`. I'm guessing you're not closing a CacheManger instance(s) or the agroal connection pool(s) if you're using it directly.

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java [hunk 1]
```diff
@@ -294,18 +298,36 @@ protected String initSelectIdRowSql() {
...
+      return "SELECT COUNT(*) FROM " + tableName;
+   }
+
    @Override
    public String getCountNonExpiredRowsSql() {
+      return countNonExpiredRowsSql;
+   }
+
+   @Override
+   public String getCountRowsSql() {
```
Comment (ryanemerson) [inline persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java:330]: Ideally we shouldn't be adding SQL to the `src/main` files just for use in test cases. This SQL is part of the standard, so it should run on all DBs, i.e. no need to override this SQL on a per DB level, so I think you should just use this directly in your test `TableManipulation` class.

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java [hunk 2]
```diff
@@ -294,18 +298,36 @@ protected String initSelectIdRowSql() {
...
+      return String.format("SELECT %s FROM %s WHERE %s LIKE ?", config.idColumnName(), tableName, config.idColumnName());
+   }
+
    @Override
    public String getSelectIdRowSql() {
       return selectIdRowSql;
    }
 
+   @Override
+   public String getSelectIdRowSqlWithLike() {
```
Comment (ryanemerson) [inline persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java:311]: Should also just be in the test `TableManipulation` class.

File: server/tests/pom.xml [hunk 3]
```diff
@@ -73,6 +73,18 @@
          <artifactId>h2</artifactId>
          <scope>test</scope>
       </dependency>
+      <dependency>
+         <groupId>mysql</groupId>
+         <artifactId>mysql-connector-java</artifactId>
+         <version>8.0.20</version>
```
Comment (ryanemerson) [inline server/tests/pom.xml:79]: We should define this version in the root `pom.xml` like we do for `h2`, e.g. `<versionx.com.h2database.h2/>`

File: server/tests/pom.xml [hunk 4]
```diff
@@ -73,6 +73,18 @@
...
+         <scope>test</scope>
+      </dependency>
+      <dependency>
+         <groupId>org.postgresql</groupId>
+         <artifactId>postgresql</artifactId>
+         <version>42.2.5</version>
```
Comment (ryanemerson) [inline server/tests/pom.xml:85]: We should define this version in the root pom.xml like we do for h2, e.g. <versionx.com.h2database.h2/>

File: server/tests/src/test/java/org/infinispan/server/persistence/TableManipulation.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+   private PooledConnectionFactoryConfigurationBuilder persistenceConfiguration;
+   private ConfigurationBuilder configurationBuilder;
+
+   public TableManipulation(String cacheName, PooledConnectionFactoryConfigurationBuilder persistenceConfiguration, ConfigurationBuilder configurationBuilder) {
+      this.cacheName = cacheName;
+      this.persistenceConfiguration = persistenceConfiguration;
+      this.configurationBuilder = configurationBuilder;
+   }
+
+   private ConnectionFactory getConnectionFactory() {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/TableManipulation.java:33]: You need to make sure the resources returned by `TableManipulation` are always closed after a test has finished ... this is probably the cause of the `ThreadLeakChecker` failures.

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java (hunk 1 above)
Comment (gustavolira) [reply persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java]: @ryanemerson updated

File: server/tests/pom.xml (hunk 4 above)
Comment (gustavolira) [reply server/tests/pom.xml]: done

File: server/tests/src/test/java/org/infinispan/server/persistence/TableManipulation.java (hunk 5 above)
Comment (gustavolira) [reply server/tests/src/test/java/org/infinispan/server/persistence/TableManipulation.java:46]: @ryanemerson I fixed some `ThreadLeakChecker` but there's one ThreadLeak that is still happening, I don't know why it's happening.

Comment (ryanemerson) [reply server/tests/src/test/java/org/infinispan/server/persistence/TableManipulation.java:46]: @wburns Can you remember what was leaking the `agroal` connection on your PR before?

Comment (wburns) [reply server/tests/src/test/java/org/infinispan/server/persistence/TableManipulation.java:46]: Mine was a bug in the start method of the PersistenceManagerImpl, I highly doubt it is related and it shouldn't be able to resurface now.

https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java#L198

Comment (gustavolira) [reply server/tests/src/test/java/org/infinispan/server/persistence/TableManipulation.java:42]: @ryanemerson I removed the dependency from `TableManager` and fixed all thread leaker

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java [hunk 6]
```diff
@@ -294,6 +294,10 @@ protected String initSelectIdRowSql() {
       return String.format("SELECT %s FROM %s WHERE %s = ?", config.idColumnName(), tableName, config.idColumnName());
    }
 
+   protected String initSelectIdRowSqlWithLike() {
```
Comment (ryanemerson) [inline persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java:297]: This method is never used

File: persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java [hunk 7]
```diff
@@ -72,7 +72,7 @@
       this.selectIdRowSql = initSelectIdRowSql();
       this.deleteRowSql = initDeleteRowSql();
       this.loadAllRowsSql = initLoadAllRowsSql();
-      this.countRowsSql = initCountNonExpiredRowsSql();
+      this.countNonExpiredRowsSql = initCountNonExpiredRowsSql();
```
Comment (ryanemerson) [inline persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/impl/table/AbstractTableManager.java:75]: As this will be the only change to the production files, I think we should revert it.

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcConfigurationUtil.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    private PooledConnectionFactoryConfigurationBuilder persistenceConfiguration;
+    private ConfigurationBuilder configurationBuilder;
+    private CacheMode cacheMode;
+
+    public JdbcConfigurationUtil(CacheMode cacheMode) {
+        configurationBuilder = new ConfigurationBuilder();
+        this.cacheMode = cacheMode;
+    }
+
+    public JdbcConfigurationUtil createPersistenceConfiguration(Database database, boolean passivation, boolean preload) {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcConfigurationUtil.java:21]: This method is always called after directly after the instance is created, e.g. `new JdbcConfigurationUtil().createPersistenceConfiguration()` ... why not just add these params to the constructor so that we can just do:

`new JdbcConfigurationUtil(CacheMode.REPL_SYNC, database, false, false).setEviction()` etc.

File: server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRule.java [hunk 9]
```diff
@@ -78,8 +78,8 @@ public void evaluate() throws Throwable {
             } finally {
                InfinispanServerRule.this.after(testName);
                if (manageServer && testServer.isDriverInitialized()) {
-                  testServer.afterListeners();
                   testServer.getDriver().stop(testName);
+                  testServer.afterListeners();
```
Comment (ryanemerson) [inline server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRule.java:82]: Why is this change required?

File: server/tests/src/test/java/org/infinispan/server/functional/MemcachedOperations.java [hunk 10]
```diff
@@ -28,4 +33,37 @@ public void testMemcachedOperations() {
       cache.set("k1", 0, "v1");
       assertEquals("v1", cache.get("k1"));
    }
+
+   @Test
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/functional/MemcachedOperations.java:37]: These changes seem fine, but do they belong in this commit? They aren't related to JDBC.

File: server/tests/src/test/java/org/infinispan/server/functional/RestOperations.java [hunk 11]
```diff
@@ -73,6 +74,19 @@ public void testRestOperations() {
...
+   @Test
+   public void testPutWithTimeToLive() throws InterruptedException {
+      RestClientConfigurationBuilder builder = new RestClientConfigurationBuilder();
+      builder.protocol(protocol);
+      RestClient client = SERVER_TEST.rest().withClientConfiguration(builder).create();
+      RestCacheClient cache = client.cache(SERVER_TEST.getMethodName());
+      sync(cache.post("k1", "v1", 1, 1));
+      assertEquals(HttpResponseStatus.OK.code(), sync(cache.get("k1")).getStatus());
+      Thread.sleep(2000);
+      assertEquals(HttpResponseStatus.NOT_FOUND.code(), sync(cache.get("k1")).getStatus());
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/functional/RestOperations.java:86]: Again this seems fine, but it isn't related to the current commit and Jira ticket. If you want to add this test, create a new Jira and add as a separate commit please.

File: server/tests/src/test/java/org/infinispan/server/persistence/AsyncJdbcStringBasedCacheStore.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        String valuePrefix = "testPutRemove-k-";
+
+        JdbcConfigurationUtil jdbcUtil = new JdbcConfigurationUtil(CacheMode.REPL_SYNC)
+                .createPersistenceConfiguration(database, false, true)
+                .setLockingConfigurations();
+        RemoteCache<String, String> cache = SERVER_TEST.hotrod().withServerConfiguration(jdbcUtil.getConfigurationBuilder()).create();
+        TableManipulation table = new TableManipulation(cache.getName(), jdbcUtil.getPersistenceConfiguration(), jdbcUtil.getConfigurationBuilder());
+
+        // test PUT operation
+        for (int i = 0; i != numEntries; i++) {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/AsyncJdbcStringBasedCacheStore.java:62]: `<` ... same for the other for loops.

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivationDisabled.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+        cache.put("k2", "v2");
+
+        assertNotNull(table.getValueByKey("k1"));
+        assertNotNull(table.getValueByKey("k2"));
+
+        SERVERS.getServerDriver().stop(0);
+        SERVERS.getServerDriver().restart(0);
+
+        // test preload==true, entries should be immediately in the cache after restart
+        assertEquals(2, cache.size());
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivationDisabled.java:109]: This isn't testing that preload works, as it is just checking the cache size which will include the loaders. Instead you should assert that the entry is in-memory by doing `cache.withFlags(Flag.SKIP_CACHE_LOAD).size()`

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivationDisabled.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        // test passivation==false, database should contain all entries which are in the cache
+        assertNotNull(table.getValueByKey("k1"));
+        assertNotNull(table.getValueByKey("k2"));
+
+        SERVERS.getServerDriver().stop(0);
+        SERVERS.getServerDriver().restart(0);
+
+        assertNotNull(table.getValueByKey("k1"));
+        assertNotNull(table.getValueByKey("k2"));
+        assertEquals("v1", cache.get("k1"));
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivationDisabled.java:80]: It's probably worth ensuring that the key is not in-memory first, so `assertNull(cache.withFlags(Flag.SKIP_CACHE_LOAD).get("k1")` followed by `assertEquals("v1", cache.get("k1"));`

File: server/tests/src/test/java/org/infinispan/server/persistence/TableManipulation.java [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+import org.infinispan.persistence.jdbc.JdbcUtil;
+import org.infinispan.persistence.jdbc.configuration.PooledConnectionFactoryConfiguration;
+import org.infinispan.persistence.jdbc.configuration.PooledConnectionFactoryConfigurationBuilder;
+import org.infinispan.persistence.jdbc.connectionfactory.ConnectionFactory;
+import org.infinispan.persistence.jdbc.impl.connectionfactory.PooledConnectionFactory;
+import org.infinispan.persistence.keymappers.DefaultTwoWayKey2StringMapper;
+
+import com.mysql.cj.jdbc.AbandonedConnectionCleanupThread;
+
+public class TableManipulation {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/TableManipulation.java:23]: If we adapt this to `implements AutoCloseable`, we can move the `releaseConnection` logic to the `stop()` implementation and then just do the following when using the code:

```java
try (TableManipulation tm = new TableManipulation(..)) {
 // Test logic
}
```

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivationEnabled.java [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+        assertEquals(1, table.countAllRows());
+        SERVERS.getServerDriver().kill(0);
+        SERVERS.getServerDriver().restart(0);
+        assertEquals(0, getNumberOfEntriesInMemory(cache.getName()));
+        assertEquals(1, table.countAllRows());
+        assertEquals("v1", cache.get("k1"));
+        table.releaseConnection();
+    }
+
+    private int getNumberOfEntriesInMemory(String cacheName) throws JsonProcessingException {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivationEnabled.java:150]: There's no need todo this via REST. We can just utilise `Flag.SKIP_CACHE_LOAD`.

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivationEnabled.java [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+    public static Collection<Object[]> data() {
+        String[] databaseTypes = PersistenceIT.DATABASE.getDatabaseTypes();
+        List<Object[]> params = new ArrayList<>(databaseTypes.length);
+        for (String databaseType : databaseTypes) {
+            params.add(new Object[]{databaseType});
+        }
+        return params;
+    }
+
+    public JdbcStringBasedCacheStorePassivationEnabled(String databaseType) {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivationEnabled.java:65]: I think we should consolidate the two `JdbcStringBasedCacheStorePassivation*` tests into a single `JdbcStringBasedCacheStore` test. We can then just use the method names to distinguish the preload enabled/disabled tests.

File: server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRule.java (hunk 9 above)
Comment (gustavolira) [reply server/testdriver/junit4/src/main/java/org/infinispan/server/test/junit4/InfinispanServerRule.java:82]: @ryanemerson Without this change we will receive `SocketTimeoutException: connect timed out` a lot

File: server/tests/src/test/java/org/infinispan/server/persistence/AsyncJdbcStringBasedCacheStore.java [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+                .setLockingConfigurations();
+        RemoteCache<String, String> cache = SERVER_TEST.hotrod().withServerConfiguration(jdbcUtil.getConfigurationBuilder()).create();
+        try(TableManipulation table = new TableManipulation(cache.getName(), jdbcUtil.getPersistenceConfiguration(), jdbcUtil.getConfigurationBuilder())) {
+            // test PUT operation
+            for (int i = 0; i < numEntries; i++) {
+                cache.putAsync(keyPrefix+i, valuePrefix+i).toCompletableFuture().get();
+            }
+
+            assertCountRow(table.countAllRows(), numEntries);
+            for (int i = 0; i != numEntries; i++) {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/AsyncJdbcStringBasedCacheStore.java:65]: <

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+            SERVERS.getServerDriver().stop(0);
+            SERVERS.getServerDriver().restart(0);
+
+            // test preload==true, entries should be immediately in the cache after restart
+            assertEquals(2, cache.size());
+            // test purge==false, entries should remain in the database after restart
+            assertNotNull(table.getValueByKey("k1"));
+            assertNotNull(table.getValueByKey("k2"));
+        }
+;
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java:121]: This is failing checkstyle.

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+            assertNull(cache.withFlags(Flag.SKIP_CACHE_LOAD).get("k3"));
+            assertEquals("v1", cache.get("k1"));
+            assertEquals("v2", cache.get("k2"));
+            //when the entry is removed from the cache, it should be also removed from the cache store (the store
+            //and the cache are the same sets of keys)
+            cache.remove("k1");
+            assertNull(table.getValueByKey("k1"));
+            cache.clear();
+        }
+
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java:96]: Lots of empty lines where you've added the `try() {}` blocks, can you remove please.

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+            assertTrue(table.countAllRows() == 0);
+            cache.put(doubleKey, doubleValue);
+            // test passivation==false, database should contain all entries which are in the cache
+            assertEquals(1, table.countAllRows());
+            assertEquals(doubleValue, cache.get(doubleKey));
+        }
+    }
+
+    @Test(timeout = 600000)
+    public void testPassivateAfterEvictionPassivationEnabled() throws Exception {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java:147]: This test is identical to the contents of `testSoftRestartWithoutPreloadPassivationEnabled` AFAICT, except it does not have the restart. `testSoftRestartWithoutPreloadPassivationEnabled` should be sufficient`

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+        }
+;
+    }
+
+    /*
+     * This should verify that DefaultTwoWayKey2StringMapper on server side can work with ByteArrayKey which
+     * is always produced by HotRod client regardless of type of key being stored in a cache.
+     */
+    @Test(timeout = 600000)
+    public void testStoreDataWithHotRodClientPassivationFalse() throws Exception {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java:129]: `testDefaultTwoWayKey2StringMapper`? That way the test name reflects what we're trying to test.

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+            //and the cache are the same sets of keys)
+            cache.remove("k1");
+            assertNull(table.getValueByKey("k1"));
+            cache.clear();
+        }
+
+    }
+
+    @Test(timeout = 600000)
+    public void testPreloadWithoutPurgePassivationFalse() throws Exception {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java:100]: The default behaviour is for `purge = false; passivation = false`, so I think `testPreload()` is sufficient.

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+        }
+        return params;
+    }
+
+    public JdbcStringBasedCacheStorePassivation(String databaseType) {
+        this.database = PersistenceIT.DATABASE.getDatabase(databaseType);
+    }
+
+    @Test(timeout = 600000)
+    public void testFailoverWithPassivationDisabled() throws Exception {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java:71]: `testFailover()`

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+            ///now some key is evicted and stored in store
+            assertTrue(2 == getNumberOfEntriesInMemory(cache.getName()));
+            assertEquals(1, table.countAllRows());
+            assertNotNull(table.getValueByKey("k1"));
+        }
+
+    }
+
+    @Test(timeout = 600000)
+    public void testSoftRestartWithoutPreloadPassivationEnabled() throws Exception {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java:169]: `testSoftRestartWithPassivation`

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java [hunk 26]
```diff
@@ -0,0 +1,234 @@
...
+
+    }
+
+    /**
+     * This test differs from the preceding expecting 1 entry in the DB
+     * after fail-over instead of 3 when doing soft
+     * restart.
+     */
+    @Test(timeout = 600000)
+    public void testFailoverWithoutPreloadPassivationEnabled() throws Exception {
```
Comment (ryanemerson) [inline server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java:202]: `testFailoverWithPassivation`

File: server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java (hunk 24 above)
Comment (wburns) [reply server/tests/src/test/java/org/infinispan/server/persistence/JdbcStringBasedCacheStorePassivation.java:71]: Btw we should not have a timeout on any test, testng listener will handle that.

**Comment by ryanemerson** [comment]
Thanks @gustavolira
